### PR TITLE
Leave websocket transport open if receive times out or is cancelled

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Leave websocket transport open if receive times out or is cancelled.

This change restores the behavior before #7978

See https://github.com/aio-libs/aiohttp/pull/7978#issuecomment-2015714011

## Are there changes in behavior for the user?

The connection is left open if the receive call times out or is cancelled

## Is it a substantial burden for the maintainers to support this?

no

## Related issue number

https://github.com/aio-libs/aiohttp/pull/7978#issuecomment-2015714011

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [ ] Add a new news fragment into the `CHANGES/` folder
  * name it `<issue_or_pr_num>.<type>.rst` (e.g. `588.bugfix.rst`)
  * if you don't have an issue number, change it to the pull request
    number after creating the PR
    * `.bugfix`: A bug fix for something the maintainers deemed an
      improper undesired behavior that got corrected to match
      pre-agreed expectations.
    * `.feature`: A new behavior, public APIs. That sort of stuff.
    * `.deprecation`: A declaration of future API removals and breaking
      changes in behavior.
    * `.breaking`: When something public is removed in a breaking way.
      Could be deprecated in an earlier release.
    * `.doc`: Notable updates to the documentation structure or build
      process.
    * `.packaging`: Notes for downstreams about unobvious side effects
      and tooling. Changes in the test invocation considerations and
      runtime assumptions.
    * `.contrib`: Stuff that affects the contributor experience. e.g.
      Running tests, building the docs, setting up the development
      environment.
    * `.misc`: Changes that are hard to assign to any of the above
      categories.
  * Make sure to use full sentences with correct case and punctuation,
    for example:
    ```rst
    Fixed issue with non-ascii contents in doctest text files
    -- by :user:`contributor-gh-handle`.
    ```

    Use the past tense or the present tense a non-imperative mood,
    referring to what's changed compared to the last released version
    of this project.

### CODE REVIEW

File: tests/test_web_websocket_functional.py [hunk 1]
```diff
@@ -799,3 +801,91 @@ async def ws_handler(request):
...
+        closed.set_result(None)
+        return ws
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    client = await aiohttp_client(app)
+
+    ws = await client.ws_connect("/", autoping=False)
+
+    await asyncio.sleep(0.1)
```
Comment (bdraco) [inline tests/test_web_websocket_functional.py]: Sleep probably doesn't make sense here, better to provide a future and resolve it once the server times out

**Comment by bdraco** [comment]
Tested on production. No adverse effects observed. Old behavior restored without the leak

**Comment by Dreamsorcerer** [comment]
And this still fixes the leak from before? I was thinking that we'd close the connection if the connection timed out (indicating that we may have lost the connection), but would keep the connection if it was cancelled externally.

**Comment by Dreamsorcerer** [comment]
Also, setting the close code to abnormal closure still seems like we're attempting to close the connection to me. That still doesn't seem like the right behaviour when the task has been cancelled externally. I think no action should be taken in that situation (though, if someone is always putting a timeout on, then maybe there are issues with it never detecting the connection is timed out?).

**Comment by bdraco** [comment]
> And this still fixes the leak from before? I was thinking that we'd close the connection if the connection timed out (indicating that we may have lost the connection), but would keep the connection if it was cancelled externally.

Yes, the original problem was in `close`, and the change to `receive` wasn't needed. I changed it to be consistent but that seems incorrect in hindsight.

**Comment by bdraco** [comment]
> Also, setting the close code to abnormal closure still seems like we're attempting to close the connection to me. That still doesn't seem like the right behaviour when the task has been cancelled externally. I think no action should be taken in that situation (though, if someone is always putting a timeout on, then maybe there are issues with it never detecting the connection is timed out?).

I did a strait revert to the previous version, but I agree that seems wrong as well.  I think it should simply raise here instead

**Comment by bdraco** [comment]
I made that change in the code, but I'm too tired to retest it so marked it back as draft until I've had more sleep

**Comment by bdraco** [comment]
Functional testing looks good on production

**Review by Dreamsorcerer — APPROVED** [review]

**Comment by bdraco** [comment]
Everything ran ok overnight. Nothing in the logs. All looks good

**Pull request merged**